### PR TITLE
Add bam_split_by_region subwf

### LINKED ISSUE #2713: new subworkflow: BAM/SPLIT_BY_REGION/SAMTOOLS
### Is there an existing subworkflow for this?

- [X] I have searched for the existing subworkflow

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this subworkflow, I added myself to the `Assignees` to facilitate tracking who is working on the subworkflow

### PR DESCRIPTION
**Pull request by TCLamnidis**
This PR adds a subworkflow to split an input bam (with a bai) to multiple bams containing the reads mapping to each region specified in an input bed/tsv file. 

This operation can be useful to increase parallelisation in various applications.

## PR checklist

Closes #2713  

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by maxulysse** [comment]
Can we use an intervals bed file to define the regions to split?

**Comment by TCLamnidis** [comment]
Not here. One could provide a bed file as a filter in the `samtools view` module `ext.args`, but not keep the reads that are not overlapping those regions. 
I am happy to make a subwf for that too if you want, but wouldn't that risk creating 0-read bam files in fringe cases? 🤔

**Comment by SPPearce** [comment]
I think this is a very useful addition.
I would ask why it needs the bam file to have an index though, I don't think you use it.

**Comment by TCLamnidis** [comment]
The index is required by the `samtools idxstats` module.

**Comment by maxulysse** [comment]
We use directly the fasta_fai in sarek to create intervals when none are supplied.
I'd rather the subworkflow did not have this part:
https://github.com/nf-core/modules/blob/eb0f25f7de1f18963de2941be741e793117e6ed5/subworkflows/nf-core/bam_split_by_chrom/main.nf#L18-L44
That way, it'd be more easy to reuse in several occasion, the creation of the chromosomes (or intervals) should be separated than the splitting of the bam file in my opinion

**Review by maxulysse — APPROVED** [review]
<3

File: subworkflows/nf-core/bam_split_by_region/main.nf [hunk 1]
```diff
@@ -0,0 +1,64 @@
+//
+// Run SAMtools idxstats, then split reads in input bam into one bam per chromosome in idxstats output
+//
+
+include { SAMTOOLS_VIEW          } from '../../../modules/nf-core/samtools/view/main'
+include { SAMTOOLS_INDEX         } from '../../../modules/nf-core/samtools/index/main'
```
Comment (maxulysse) [inline subworkflows/nf-core/bam_split_by_region/main.nf:6]:
```suggestion
include { SAMTOOLS_VIEW  } from '../../../modules/nf-core/samtools/view/main'
include { SAMTOOLS_INDEX } from '../../../modules/nf-core/samtools/index/main'
```

File: subworkflows/nf-core/bam_split_by_region/main.nf [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                                region = stats['seq_name']
+                            } else if ( ! stats['stop']) {
+                            // If a specific position is given, use that
+                                region = [stats['seq_name'], stats['start']].join(":")
+                            } else {
+                            // If a region between specific bps is requested use that
+                                region = [ [stats['seq_name'], stats['start']].join(":"), stats['stop']].join('-')
+                            }
+                        [ region ]
+                    }
```
Comment (maxulysse) [inline subworkflows/nf-core/bam_split_by_region/main.nf:34]: .map{ stats ->
                        // If the regions file contains just a sequence name provide that
                        if (! stats['start'] ) [ stats['seq_name'] ]
                        // If a specific position is given, use that
                        else if ( ! stats['stop']) [ [ stats['seq_name'], stats['start'] ].join(":") ]
                        // If a region between specific bps is requested use that
                        else [ [ [ stats['seq_name'], stats['start'] ].join(":"), stats['stop'] ].join('-') ]
                    }

File: subworkflows/nf-core/bam_split_by_region/main.nf [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    ch_bam
+        .combine(ch_regions)
+        // Place region into map
+        .map{
+            meta, bam, bai, chrom ->
+            clone = meta.clone()
+            clone['genomic_region'] = chrom
+
+            [ clone, bam, bai ]
+        }
```
Comment (maxulysse) [inline subworkflows/nf-core/bam_split_by_region/main.nf:46]:
```suggestion
        .map{ meta, bam, bai, chrom -> [ meta + [ genomic_region:chrom ], bam, bai ] }
```

File: subworkflows/nf-core/bam_split_by_region/main.nf (hunk 2 above)
Comment (maxulysse) [reply subworkflows/nf-core/bam_split_by_region/main.nf:34]:
```suggestion
                .map{ stats ->
                    // If the regions file contains just a sequence name provide that
                    if (! stats['start'] ) [ stats['seq_name'] ]
                    // If a specific position is given, use that
                    else if ( ! stats['stop']) [ [ stats['seq_name'], stats['start'] ].join(":") ]
                    // If a region between specific bps is requested use that
                    else [ [ [ stats['seq_name'], stats['start'] ].join(":"), stats['stop'] ].join('-') ]
                }
```

**Comment by maxulysse** [comment]
made a couple of code suggestions, but looks amazing.
I absolutely love this

File: subworkflows/nf-core/bam_split_by_region/main.nf (hunk 3 above)
Comment (TCLamnidis) [reply subworkflows/nf-core/bam_split_by_region/main.nf:46]: oooooh, that's some fancy syntax! <3

**Comment by TCLamnidis** [comment]
eager. In the dsl2 rewrite, I want to do deduplication per chromosome to avoid pesky java memory issues due to large datasets

**Pull request merged**